### ISSUE TITLE
rfc, flat transaction method ?

### ISSUE THREAD

**Issue by SignpostMarv** [paragonie/easydb#42]
A pattern I'm using a lot at work is something like this:

```php
$db = $cms->getDatabaseConnection();
$autoStartTransaction = ($db->inTransaction() === false);
if ($autoStartTransaction) {
    $db->beginTransaction();
}
try {
/**
Do thing here
*/
    if ($autoStartTransaction) {
        $db->commit();
    }
} catch (Throwable $e) {
    if ($autoStartTransaction) {
        $db->rollBack();
    }
    throw $e;
}
```

I'm wondering if there'd be a desirable utility in having a method with a signature something like this:

```php
public function TryFlatTransaction(callable $cb)
{
    $autoStartTransaction = ($db->inTransaction() === false);
    if ($autoStartTransaction) {
        $db->beginTransaction();
    }
    try {
        $cb($this);
        if ($autoStartTransaction) {
            $db->commit();
        }
    } catch (Throwable $e) {
        if ($autoStartTransaction) {
            $db->rollBack();
        }
        throw $e;
    }
}
```

Where usage is something like this:
```php
class Foo
{
    private $db;
    function __construct(EasyDB $db) {
        $this->db = $db;
    }

    function doThing($input) {
        $this->db->TryFlatTransaction(function (EasyDB $db) use ($input) {
            $db->safeQuery(/* some query */, /* some args from $input*/); // might throw an exception
            /**
            do something else that might throw an exception
            */
        });
    }

    function doRecursiveThing($input) {
        foreach ($input as $maybeRecursive) {
            $this->db->TryFlatTransaction(function (EasyDB $db) use ($maybeRecursive) {
                $this->doThing($maybeRecursive);
                if (
                    // some recursive check
                ) {
                    $this->doRecursiveThing($maybeRecursive);
                }
            });
        }
    }
}
```

**Comment by paragonie-scott** [comment]
This is implemented with the merging of #44. Thanks @SignpostMarv :)

**Comment by SignpostMarv** [comment]
no probs :)
